Add tracing to Ray (#14872)
ray-project/ray: doc/source/ray-tracing.rst

Users who want to experiment with tracing can configure their remote span processors to export spans to a local JSON file. Serious users developing locally can push their traces to Jaeger containers via the `Jaeger exporter <https://open-telemetry.github.io/opentelemetry-python/exporter/jaeger/jaeger.html>`_.

Additional Instruments
~~~~~~~~~~~~~~~~~~~~~~
If you are using a library that has built-in tracing support, the ``setup_tracing`` function you provide should also patch those libraries. You can find more documentation for the instrumentation of these libraries `here <https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation>`_.

Below is an example tracing startup hook that sets up the default tracing provider, exports spans to files in ``/tmp/spans``, and does not have any additional instruments.

.. code-block:: python

  import ray
  import os
  from opentelemetry import trace
  from opentelemetry.sdk.trace import TracerProvider
  from opentelemetry.sdk.trace.export import (
      ConsoleSpanExporter,
      SimpleExportSpanProcessor,
  )
  
  

>   def setup_tracing() -> None:

Reviewer: This is a question and not a suggestion, but I'm still having trouble wrapping my head around what this setup function is supposed to do.

There's some gap in my understanding, but I'm imagining that Ray is generating events (e.g., tuples like `('event-name', start_timestamp, end_timestamp)` or something) and we want those events to get sent somewhere (e.g., a file or a service or something).

So I would imagine that the information the user needs to provide is
1. Tell Ray where to send the events
2. Optionally what format to put the events in (but maybe everything is standardized and this is unnecessary)

Given that, I would expect the configuration to either look like
1. The address of the place to write the events.
2. A function that takes in an event, formats it, and writes it to the correct place.

The code here doesn't really seem to fit either of those descriptions.

Author: Ah I see what you are saying! Tracing can be thought of as a different process than metrics (where metrics would be more similar to gathering information outputted by Ray).  You can think of tracing as simply wrapping different events and gathering all the necessary information of timestamps, event names, etc. automatically. So if you wrap a code sample with `with tracer.start_as_current_span("foo"):`, the name of the span would be `foo` and it would automatically track the timestamps and such! Does that make sense? Rather than collecting events after they have occurred, tracing is gathering the information *as the events occur.*

Reviewer 2: Additionally, users are not telling Ray where to send the events, they are telling OpenTelemetry where to send it. Ray just push events to OpenTelemetry functions and those functions will send out the events to proper destination. 

This segment encodes "where to send it": 
```
trace.get_tracer_provider().add_span_processor(
          SimpleExportSpanProcessor(
              ConsoleSpanExporter(
                  out=open(f"/tmp/spans/{os.getpid()}.json", "a")
                  )
          )
      )
```

Reviewer 3: Yeah the interesting point here is that there's nothing specific to tracing about the interface; we discussed offline and it seems to make sense to turn this into a general startup hook for worker processes that could be more widely useful.

Additional Instruments
~~~~~~~~~~~~~~~~~~~~~~
If you are using a library that has built-in tracing support, the ``setup_tracing`` function you provide should also patch those libraries. You can find more documentation for the instrumentation of these libraries `here <https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation>`_.

Below is an example tracing startup hook that sets up the default tracing provider, exports spans to files in ``/tmp/spans``, and does not have any additional instruments.
 
.. code-block:: python

  import ray
  import os
  from opentelemetry import trace
  from opentelemetry.sdk.trace import TracerProvider
  from opentelemetry.sdk.trace.export import (
      ConsoleSpanExporter,
      SimpleExportSpanProcessor,
  )
  
  
  def setup_tracing() -> None:
      # Sets the tracer_provider. This is only allowed once per execution

>       # context and will log a warning if attempted multiple times.

Reviewer 2: os.makedirs here?

      trace.set_tracer_provider(TracerProvider())
      trace.get_tracer_provider().add_span_processor(
          SimpleExportSpanProcessor(
              ConsoleSpanExporter(
                  out=open(f"/tmp/spans/{os.getpid()}.json", "a")
                  )
          )
      )

For open-source users who want to experiment with tracing, Ray has a default tracing startup hook that exports spans to the folder ``/tmp/spans``. To run using this default hook, you can run the following code sample to set up tracing and trace a simple Ray task.

.. tabs::
    .. group-tab:: ray start

        .. code-block:: shell

>             $ ray start --head --tracing-startup-hook=ray.util.tracing.enable_tracing:setup_tracing

Reviewer 2: make sure to change the utils enable_tracing to tests.enable tracing to convey that this is really just used for proof of concept not for production.

Author: done

Users who want to experiment with tracing can configure their remote span processors to export spans to a local JSON file. Serious users developing locally can push their traces to Jaeger containers via the `Jaeger exporter <https://open-telemetry.github.io/opentelemetry-python/exporter/jaeger/jaeger.html>`_.

Additional Instruments
~~~~~~~~~~~~~~~~~~~~~~
If you are using a library that has built-in tracing support, the ``setup_tracing`` function you provide should also patch those libraries. You can find more documentation for the instrumentation of these libraries `here <https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation>`_.

Below is an example tracing startup hook that sets up the default tracing provider, exports spans to files in ``/tmp/spans``, and does not have any additional instruments.

.. code-block:: python

  import ray
  import os
  from opentelemetry import trace
  from opentelemetry.sdk.trace import TracerProvider
  from opentelemetry.sdk.trace.export import (
      ConsoleSpanExporter,
      SimpleExportSpanProcessor,
  )
  from typing import Any
  
  

>   def setup_tracing(*args: Any, **kwargs: Any) -> None:

Reviewer 2: this should take 0 arguments, since in our hook (worker.py) we call it directly without any args

Author: changed

      # Sets the tracer_provider. This is only allowed once per execution
      # context and will log a warning if attempted multiple times.
      trace.set_tracer_provider(TracerProvider())
      trace.get_tracer_provider().add_span_processor(
          SimpleExportSpanProcessor(
              ConsoleSpanExporter(
                  out=open(f"/tmp/spans_file{os.getpid()}.json", "a")
                  )
          )
      )

    
To run your program with the tracing hook, see the following examples.

.. tabs::
  .. code-tab:: bash

    $ ray start --head --tracing-startup-hook "my_library:setup_tracing"
    $ python

>     >>> ray.init(address="auto")

Reviewer 2: add some ray tasks invocation so the examples will output some traces.

  .. code-tab:: python

    ray.init(_tracing_startup_hook="my_library:setup_tracing")

> 

Reviewer 2: maybe add a note with the setup_tracing we have in our tracing so people can play with it out of the box?

Tracing
=======

> To help debug and monitor Ray applications, we have built in tracing into Ray libraries. 

Reviewer 3: ```suggestion
To help debug and monitor Ray applications, Ray integrates with OpenTelemetry to make it easy to export traces to external tracing stacks such as Jaeger. 
```

.. note::

    Tracing is currently an experimental feature and under active development. APIs are subject to change.

Getting Started
---------------

> First, install opentelemetry.

Reviewer 3: ```suggestion
First, install OpenTelemetry.
```

.. code-block:: bash

    pip install opentelemetry-api==1.0.0rc1
    pip install opentelemetry-sdk==1.0.0rc1

To enable tracing, you must provide a tracing startup hook with a function that will set up the Tracer Provider, Remote Span Processors, and Additional Instruments. 

Tracer Provider
~~~~~~~~~~~~~~~~
This configures how to collect traces. View the TracerProvider API `here <https://open-telemetry.github.io/opentelemetry-python/sdk/trace.html#opentelemetry.sdk.trace.TracerProvider>`__.

Remote Span Processors
~~~~~~~~~~~~~~~~~~~~~~

> This configures where to export traces. View the SpanProcessor API `here <https://open-telemetry.github.io/opentelemetry-python/sdk/trace.html#opentelemetry.sdk.trace.SpanProcessor>`__.

Reviewer 3: ```suggestion
This configures where to export traces to. View the SpanProcessor API `here <https://open-telemetry.github.io/opentelemetry-python/sdk/trace.html#opentelemetry.sdk.trace.SpanProcessor>`__.
```

Users who want to experiment with tracing can configure their remote span processors to export spans to a local JSON file. Serious users developing locally can push their traces to Jaeger containers via the `Jaeger exporter <https://open-telemetry.github.io/opentelemetry-python/exporter/jaeger/jaeger.html>`_.

Additional Instruments
~~~~~~~~~~~~~~~~~~~~~~

> If you are using a library that has built-in tracing support, the setup_tracing function you provide should also patch those libraries. You can find more documentation for the instrumentation of these libraries `here <https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation>`_.

Reviewer 3: ```suggestion
If you are using a library that has built-in tracing support, the ``setup_tracing`` function you provide should also patch those libraries. You can find more documentation for the instrumentation of these libraries `here <https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation>`_.
```

> Below is an example tracing startup hook that sets up the default Tracing Provider, exports spans to files in /tmp/spans, and does not have any Additional Instruments.

Reviewer 3: ```suggestion
Below is an example tracing startup hook that sets up the default tracing provider, exports spans to files in ``/tmp/spans``, and does not have any additional instruments.
```

.. code-block:: python

  import ray
  import os
  from opentelemetry import trace
  from opentelemetry.sdk.trace import TracerProvider
  from opentelemetry.sdk.trace.export import (
      ConsoleSpanExporter,
      SimpleExportSpanProcessor,
  )
  from typing import Any
  
  
  def setup_tracing(*args: Any, **kwargs: Any) -> None:
      if getattr(ray, "__traced__", False):
          return
    

>       ray.__traced__ = True

Reviewer 3: I would find this very confusing as a user who wasn't deeply familiar with ray/tracing/monkey patching. Please add a comment about what's going on and why this is necessary.

Reviewer 3: Actually, is it possible to do this check automatically to ensure it's only run once per process? That way the user doesn't even need to think about it (do the same check and modification in the outer code that calls this hook).

Author: oh yeah, i actually moved it there already but forgot to update the doc, ill remove this

      # Sets the tracer_provider. This is only allowed once per execution
      # context and will log a warning if attempted multiple times.
      trace.set_tracer_provider(TracerProvider())
      trace.get_tracer_provider().add_span_processor(
          SimpleExportSpanProcessor(
              ConsoleSpanExporter(
                  out=open(f"/tmp/spans_file{os.getpid()}.json", "a")
                  )
          )
      )

    

> To run your program with the tracing hook, see the following examples.

Reviewer 3: This needs more explanation about what the tracing hook is. Explain that this hook needs to be available in the python environment of all of the worker processes and is expected to be a function that is called with no arguments or kwargs.

.. tabs::
  .. code-tab:: start

>     $ ray start --head --tracing-startup-hook "MyLibrary:setup_tracing"

Reviewer 3: ```suggestion
    $ ray start --head --tracing-startup-hook "my_library:setup_tracing"
```

    $ python
    >>> ray.init(address="auto")

  .. code-tab:: init

    $ python

>     >>> ray.init(_tracing_startup_hook="MyLibrary:setup_tracing")

Reviewer 3: ```suggestion
    >>> ray.init(_tracing_startup_hook="my_library:setup_tracing")
```

> Custom traces

Reviewer 3: ```suggestion
Custom Traces
```

*************

> You can easily add custom tracing in your programs. Within your program, get the tracer object and then call trace.get_tracer(__name__)

Reviewer 3: ```suggestion
You can easily add custom tracing in your programs. Within your program, get the tracer object and then call ``trace.get_tracer(__name__)``.
```
